pkg/session/session.go

```diff
@@ -4179,6 +4182,7 @@ func (s *session) SetMemoryFootprintChangeHook() {
 		s.sessionVars.MemDBFootprint.ReplaceBytesUsed(int64(mem))
 	}
 	s.txn.SetMemoryFootprintChangeHook(hook)
+	s.GetSessionVars().TxnCtx.MemHookAlreadySet = true
```

Comment:
Perhaps it's better to use `GetVars` or other `GetX` interfaces to get the hook from `Transaction` interface instead of putting this state into the `TxnCtx` belonging to the session vars, as the `hook` is actually set into the `transation` object.

Reply:
Okay I did a microbenchmark. Seems acceptable.
```
vanilla
118.8 ns/op	      51 B/op	       3 allocs/op

bool in session vars
2.331 ns/op	       0 B/op	       0 allocs/op

get from memdb
4.881 ns/op	       0 B/op	       0 allocs/op
```